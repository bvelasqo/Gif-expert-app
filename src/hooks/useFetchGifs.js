import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

// Hook to use fetchGifs and return the results of the API call as a state variable and a function to set the state variable (setGifs)
export const useFetchGifs = (category) => {
  const [state, setState] = useState({data: [], loading: true});

  // useEffect is a hook that runs a function when a component is rendered
  useEffect(() => {
    // getGifs is a function that returns a promise (async function)
    getGifs(category).then((images) => {
      // setState is a function that takes a state variable and a function as arguments
      setState({
        data: images,
        loading: false,
      });
    }).catch(() => {
      // setState is a function that takes a state variable and a function as arguments
      setState({
        loading: false,
      });
    });
  }, [category]);
  // [] is an array of dependencies, if the dependencies change, the function will run again (useEffect runs only once)

  // Return the state variable and the function to set the state variable (setGifs)
  return state;
}
