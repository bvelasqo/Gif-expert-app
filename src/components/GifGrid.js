import React from "react";
import PropTypes from "prop-types";
import {useFetchGifs} from "../hooks/useFetchGifs";
import {GifGridItem} from "./GifGridItem";

const GifGrid = ({category}) => {
	const {data: gifs, loading} = useFetchGifs(category);

	return (
		<div className="main_gif_grid">
			<h3 className="category animate__animated animate__bounceIn">Category: {category}</h3>
			{/* loading con && */}
			{loading && <p>Loading...</p>}
			{/* loading con || */}
			{loading ? (
				<div className="animate__animated animate__flash">
					<img
						src="https://media.giphy.com/media/3oEjI6SIIHBdRxXI40/giphy.gif"
						alt="loading"
					/>
				</div>
			) : (
				<div className="gif_grid">
					{gifs.map((image) => (
						<>
							<GifGridItem key={image.id} {...image} />
						</>
					))}
				</div>
			)}
		</div>
	);
};

GifGrid.prototype = {
	category: PropTypes.string.isRequired,
};

export default GifGrid;
