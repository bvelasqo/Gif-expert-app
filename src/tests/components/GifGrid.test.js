import "@testing-library/jest-dom";
import shallow from "enzyme/shallow";
import GifGrid from "../../components/GifGrid";
import {useFetchGifs} from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("GifGrid", () => {
	const category = "Dragon Ball";
  let gifs;
  let wrapper;
	beforeEach(() => {
		jest.clearAllMocks();
		const gifs = [
			{
				id: "ABC",
				url: "https://localhost/cualquiercosa.jpg",
				title: "Cualquier cosa",
			},
		];
		useFetchGifs.mockReturnValue({
			data: gifs,
			loading: false,
		});
	});
	test("should render correctly", () => {
		const wrapper = shallow(<GifGrid category={category} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("Should to show items when useFetchGifs is called", () => {
		const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper.find("GridItem").exists()).toBe(false);
	});

  test("Should to show loading when useFetchGifs is called", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper.find("#loading").exists()).toBe(true);
  });
});
