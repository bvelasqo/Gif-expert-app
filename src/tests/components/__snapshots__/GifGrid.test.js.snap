// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid should render correctly 1`] = `
<div
  className="main_gif_grid"
>
  <h3
    className="category animate__animated animate__bounceIn"
  >
    Category: 
    Dragon Ball
  </h3>
  <div
    className="gif_grid"
    key="2"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/cualquiercosa.jpg"
    />
  </div>
</div>
`;
